tripal/tripal
# add path alias to tripal entity base class

# New Feature

### Closes #1963
### Closes #1973
### Dependant on #2019 

### Tripal Version: 4.x

## Description
This adds a path alias field to tripal entity widgets, allowing entry of an alias when content is first created, or changing it when the entity is edited.
If an alias is not added, or if it is removed, then the default alias is generated (this default is the same as before this PR).
This PR also fixes the problem of multiple copies of the default alias being generated. Now exactly one alias should exist for every Tripal entity.

![image](https://github.com/user-attachments/assets/dbb2994d-1cf0-40f4-9326-e3d767f16935)

## Testing?
1. Create any content type, and do not specify an alias when creating it
2. Check the alias table (or go to `/admin/config/search/path`)
`select * from path_alias;`
3. Edit, the default alias should populate the alias box. Change it to a new value `/p2` and save
4. Check the alias table (or go to `/admin/config/search/path`)
`select * from path_alias;`
```
sitedb=> select * from path_alias;         
 id | revision_id |                 uuid                 | langcode |    path     |   alias    | status 
----+-------------+--------------------------------------+----------+-------------+------------+--------
  1 |           1 | ad583371-adb5-4392-9f8f-5648349f2bc5 | und      | /bio_data/2 | /p2        |      1
```
5. Edit and leave the alias box blank and save.
6. Check the alias table (or go to `/admin/config/search/path`)
`select * from path_alias;`
The alias should revert to the default alias.
![2024-09-25_aliascardinality1](https://github.com/user-attachments/assets/5f67b10d-bc44-4586-88b4-5fbfbbc0a229)
7. Edit the content type and set the URL alias pattern to include field tokens (e.g. `organism/[organism_genus]-[organism_species]`)
8. Now either edit an existing page or create a new one and leave the alias blank or delete it. Ensure that there are no errors and the tokens are filled in as expected.
7. Create a new entity and specify an alias when creating it.
